synnax/pkg/distribution/framer/deleter/lease_proxy.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+package deleter
+
+import (
+	"context"
+	"github.com/samber/lo"
+	"github.com/synnaxlabs/aspen"
+	"github.com/synnaxlabs/synnax/pkg/distribution/channel"
+	"github.com/synnaxlabs/synnax/pkg/storage/ts"
+	"github.com/synnaxlabs/x/errors"
+	"github.com/synnaxlabs/x/telem"
+
+	"github.com/synnaxlabs/synnax/pkg/distribution/core"
+	"github.com/synnaxlabs/synnax/pkg/distribution/proxy"
+)
+
+type leaseProxy struct {
+	ServiceConfig
+	keyRouter proxy.BatchFactory[channel.Key]
+}
+
+func newLeaseProxy(
+	cfg ServiceConfig,
+) (*leaseProxy, error) {
+	p := &leaseProxy{
+		ServiceConfig: cfg,
+		keyRouter:     proxy.BatchFactory[channel.Key]{Host: cfg.HostResolver.HostKey()},
+	}
+	return p, nil
+}
+
+func (lp *leaseProxy) deleteTimeRange(
+	ctx context.Context,
+	keys channel.Keys,
+	tr telem.TimeRange,
+) error {
+	batch := lp.keyRouter.Batch(keys)
+	if len(batch.Free) != 0 {
+		return errors.Newf(
+			"cannot delete time range from virtual channel(s) %s",
+			batch.Free,
+		)
+	}
+
+	for nodeKey, entries := range batch.Peers {
+		err := lp.deleteTimeRangeRemote(ctx, nodeKey, entries, tr)
+		if err != nil {
+			return err
+		}
+	}
+
+	return lp.deleteTimeRangeGateway(ctx, batch.Gateway, tr)
+}
+
+func (lp *leaseProxy) deleteTimeRangeByName(
+	ctx context.Context,
+	retriever channel.Readable,
+	names []string,
+	tr telem.TimeRange,
+) error {
+	var res []channel.Channel
```

Comment:
I would pre-allocate this array with a specific capacity equal to the length of names. Deleting by name is also kind of sus because there could be many channels with the same name

Reply:
But I guess that's just a caveat we need to note in the docs

Reply:
problematic if a name matches multiple results and one name is not found: unique names would have same length as desired names. Is the main concern for not preallocating for memory?